api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,407 @@
+		for i := 0; i < int(SyncLoopBatchSize); i++ {
+			epochLastBlock := shard.Schedule.EpochLastBlock(curEpoch.Uint64())
+			if epochLastBlock > maxHeight {
+				break
+			}
+			heights = append(heights, epochLastBlock)
+			curEpoch = curEpoch.Add(curEpoch, common.Big1)
+		}
+
+		err := ss.ProcessStateSync(heights, bc, worker)
+		if err != nil {
+			utils.Logger().Error().Err(err).
+				Msgf("[SYNC] ProcessStateSync failed (isBeacon: %t, ShardID: %d, otherHeight: %d, currentHeight: %d)",
+					isBeacon, bc.ShardID(), maxHeight, height)
+			break
+		}
+	}
+}
+
+// getMaxPeerHeight gets the maximum blockchain heights from peers
+func (ss *EpochSync) getMaxPeerHeight(isBeacon bool) uint64 {
+	maxHeight := uint64(math.MaxUint64)
+	var (
+		wg   sync.WaitGroup
+		lock sync.Mutex
+	)
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			//debug
+			// utils.Logger().Debug().Bool("isBeacon", isBeacon).Str("peerIP", peerConfig.ip).Str("peerPort", peerConfig.port).Msg("[Sync]getMaxPeerHeight")
+			response, err := peerConfig.client.GetBlockChainHeight()
+			if err != nil {
+				utils.Logger().Warn().Err(err).Str("peerIP", peerConfig.ip).Str("peerPort", peerConfig.port).Msg("[Sync]GetBlockChainHeight failed")
+				ss.syncConfig.RemovePeer(peerConfig)
+				return
+			}
+			utils.Logger().Info().Str("peerIP", peerConfig.ip).Uint64("blockHeight", response.BlockHeight).
+				Msg("[SYNC] getMaxPeerHeight")
+
+			lock.Lock()
+			if response != nil {
+				if maxHeight == uint64(math.MaxUint64) || maxHeight < response.BlockHeight {
+					maxHeight = response.BlockHeight
+				}
+			}
+			lock.Unlock()
+		}()
+		return
+	})
+	wg.Wait()
+	return maxHeight
+}
+
+// ProcessStateSync processes state sync from the blocks received but not yet processed so far
+func (ss *EpochSync) ProcessStateSync(heights []uint64, bc *core.BlockChain, worker *worker.Worker) error {
+
+	var payload [][]byte
+
+	ss.syncConfig.ForEachPeer(func(peerConfig *SyncPeerConfig) (brk bool) {
+		resp := peerConfig.GetClient().GetBlocksByHeights(heights, ss.selfip, ss.selfport)
+		if resp == nil {
+			return false
+		}
+		if len(resp.Payload) == 0 {
+			return false
+		}
+		payload = resp.Payload
+		return true
+	})
+
+	if len(payload) == 0 {
+		return errors.New("empty payload")
+	}
+
+	decoded := make([]*types.Block, 0, len(payload))
+	for idx, blockBytes := range payload {
+		block, err := RlpDecodeBlockOrBlockWithSig(blockBytes)
+		if err != nil {
+			return errors.Wrap(err, "failed decode")
+		}
+
+		if !block.IsLastBlockInEpoch() {
+			return errors.Errorf("block is not last index %d hash %s", idx, block.Hash().Hex())
+		}
+
+		decoded = append(decoded, block)
+	}
+	// TODO(k.potapov) validate blocks
+
+	for _, block := range decoded {
+		_, err := bc.StoreShardStateBytes(block.Header().Epoch(), block.Header().ShardState())
+		if err != nil {
+			return err
+		}
+
+		err = bc.WriteBlockWithoutState(block, common.Big1)
```

Comment:
Have you tested locally or on mainnet? Is there any error returned from this line?

Reply:
There is a lot of messages written to leveldb other than the state. Furthermore, it seems there are not enough check on the block message.

Reply:
Yes, it works locally, i'm going to test it on testnet. There are no checks on block inside WriteBlockWithoutState, i'll add signature validation, other impossible without state.

Reply:
Please take a look at Finalize function at `internal/chain/engine.go:276`. You will at least need the state shard for the signature check.

Reply:
Yea, i checked that function, but it seems my changes does not affect it. Shard state functions called only for beacon chain, and my code works only for non-beacon.

Reply:
You need to call WriteShardStateBytes() or something related to write the shardState to DB. Take a look at WriteBlockWithState and see how it works, especially the part about offChain data. Then you will know why WriteShardStateBytes() is needed. Because otherwise, the shard state can be read from the DB.

Reply:
I use StoreShardStateBytes above, that function is a wrapper for `WriteShardStateBytes` and calls it internally. So, yes, i store shardState to DB. `WriteBlockWithoutState` i need for height in order to understand where i am right now and what to do next. Without `WriteBlockWithoutState` it starts to panic in different places, it's better not to skip that.